drain3/template_miner.py

```diff
@@ -170,24 +171,93 @@ def match(self, log_message: str, full_search_strategy="never") -> LogCluster:
         escaped_prefix = re.escape(self.config.mask_prefix)
         escaped_suffix = re.escape(self.config.mask_suffix)
-        template_regex = re.sub(escaped_prefix + r".+?" + escaped_suffix, self.drain.param_str, log_template)
-        if self.drain.param_str not in template_regex:
-            return []
-        template_regex = re.escape(template_regex)
-        template_regex = re.sub(r'\\ +', r'\\s+', template_regex)
-        template_regex = "^" + template_regex.replace(escaped_prefix + r"\*" + escaped_suffix, "(.*?)") + "$"
+        template_regex = re.escape(log_template)
+        parameter_groups = dict()
+        named_groups = set()
+        def unique_name(original):
+            while original in named_groups:
+                original += str(random.randint(0, 100000))
+            named_groups.add(original)
+            return original
+
+        # Mask-substitution: For every mask in the template, replace it with
+        # a named group of all possible masking-patterns it could represent (in order).
+        for mask, patterns in mask_patterns.items():
+            def create_parameter_group(match):
+                modified_patterns = []
+                for pattern, groups in patterns:
+                    # Any pattern may named groups that have already been used.
+                    # Therefore the affected groups are given a new unique name.
+                    for group_name in groups:
+                        if group_name not in named_groups:
+                            continue
+                        new_name = unique_name(group_name)
+                        pattern = re.sub(r"\(\?P={}\)".format(re.escape(group_name)), lambda x: "(?P={})".format(new_name), pattern)
+                        pattern = re.sub(r"\(\?P<{}>".format(re.escape(group_name)), lambda x: "(?P<{}>".format(new_name), pattern)
+                    named_groups.update(groups)
+                    modified_patterns.append(pattern)
+                parameter_group = unique_name("parameter")
+                parameter_groups[parameter_group] = mask
+                return "(?P<{}>".format(parameter_group) + "|".join(modified_patterns) + ")"
+            template_regex = re.sub(re.escape(escaped_prefix) + re.escape(re.escape(mask)) + re.escape(escaped_suffix),
+                                    create_parameter_group, template_regex)
+
+        template_regex = re.sub(r"\\ ", r"\\s+", template_regex)
+        template_regex = "^" + template_regex + "$"
 
         for delimiter in self.config.drain_extra_delimiters:
-            content = re.sub(delimiter, ' ', content)
-        parameter_list = re.findall(template_regex, content)
-        parameter_list = parameter_list[0] if parameter_list else ()
-        parameter_list = list(parameter_list) if isinstance(parameter_list, tuple) else [parameter_list]
+            content = re.sub(delimiter, " ", content)
+        parameter_match = re.match(template_regex, content)
```

Comment:
I assume another big performance hit is the matching of the more complex pattern.
I'd say there is no clear optimization potential for this, other than extracting the parameters directly when mining/matching the message.